### ISSUE TITLE
Deploy monitoring app and make it faster to load

### ISSUE THREAD

**Issue by kongzii** [gnosis/prediction-market-agent-tooling#89]

**Comment by kongzii** [comment]
Deployed on https://prediction-market-agent-tooling-monitor.streamlit.app/, Evan is looking into making it faster

**Comment by evangriffiths** [comment]
Needs profiling to confirm, but I think it's because `DeployedManifoldAgent.get_resolved_bets` makes an API call for every bet (`get_manifold_market(bet.contractId)`).

**Comment by kongzii** [comment]
> Needs profiling to confirm, but I think it's because `DeployedManifoldAgent.get_resolved_bets` makes an API call for every bet (`get_manifold_market(bet.contractId)`).

Sounds possible, you could try to run that for loop via https://github.com/polywrap/evo.researcher/blob/3d733c544314a4ad665993abecb74b6a2bb69d2d/evo_researcher/functions/parallelism.py#L12

**Comment by evangriffiths** [comment]
Nice. Yeah, was going to try multithreading :)

**Comment by gabrielfior** [comment]
Would also suggest taking a look at asyncio parallelism instead of
concurrent.futures due to widespread usage

**Comment by evangriffiths** [comment]
<img width="977" alt="Screenshot 2024-03-13 at 11 12 44" src="https://github.com/gnosis/prediction-market-agent-tooling/assets/56087052/45632488-564c-4a9a-8e70-04813d4f76b5">

3/4 time spent getting markets corresponding to manifold bets, the other 1/4 getting markets. Both can be spread up with multithreading.

@gabrielfior 

I'm more familiar with multithreading using `concurrent.futures`. e.g. we use it here in the benchmarking code: https://github.com/gnosis/prediction-market-agent-tooling/blob/c31c346d5a8fbfcf53dadb638482f9fca01f0f97/prediction_market_agent_tooling/benchmark/benchmark.py#L164. What are the advantages of `asyncio` here?

**Comment by kongzii** [comment]
> Would also suggest taking a look at asyncio parallelism instead of
concurrent.futures due to widespread usage

I benchmarked this in my previous company. Unless we would work on some heavily used API and the whole codebase would be async, then there is no speed difference between using the par_map I posted above vs. doing it in an async way. (in situations like this, where we need to fetch 50/500/1000 requests from some API)

Actually using threads was a bit faster and more straightforward, as one can always just replace for loop with `par_map(my list, my function)` in any part of the codebase, without worrying about anyting.

**Comment by gabrielfior** [comment]
> Actually using threads was a bit faster and more straightforward, as one can always just replace for loop with `par_map(my list, my function)` in any part of the codebase, without worrying about anyting.

I was also going to mention that, for this simple use case (sending requests in parallel), letting `asyncio` manage the coroutines or simply firing them using `concurrent.futures` is equivalent - curious to discuss more about the benchmark you had @kongzii if we have the need to implement more parallelism.

My understanding of asyncio is that it offers a bit more control in how to control the coroutines, debug them and have more control over the event loop. Don't think that's extremely important from what I understand from the use case.

Fine with both options for the implementation discussed here.

**Comment by kongzii** [comment]
> and have more control over the event loop.

The last time I used coroutines heavily, there wasn't much to do about controlling them. It was like, just do the await stuff, and everything else is going on in the background. We had problems like too many coroutines being spawned and no way to limit it down, so in the end, our API was async, but at the same time, we had to spawn multiple threads 😄 

Did they improve that in the meantime?
